gio: Add Send implementation to IOStreamAsyncReadWrite

philn opened the pull request:
This is a wrapper for GSocketConnection, which should be OK to share across
threads.

philn committed:
gio: Add Send implementation to IOStreamAsyncReadWrite

This is a wrapper for GSocketConnection, which should be OK to share across
threads.

sdroege commented on src/gio.rs:
```diff
@@ -68,6 +68,8 @@ pub struct IOStreamAsyncReadWrite<T: IsA<gio::IOStream>> {
     write: gio::OutputStreamAsyncWrite<gio::PollableOutputStream>,
 }
 
+unsafe impl<T: IsA<gio::IOStream>> Send for IOStreamAsyncReadWrite<T> {}
```
Hm actually this is too wide. This would allow to have multiple references to the `gio::IOStream` before creating the wrapper.

Meaning, we need another wrapper around this for `connect_async()` where we can guarantee that nothing else has a reference to it, and then on that wrapper implement `Send`.

philn replied:
A wrapper around `IOStreamAsyncReadWrite` ?

sdroege replied:
Yes, for usage inside `connect_async()`

philn replied:
TBH I'm not sure to understand where the issue is :) I'll push a new version adding that new wrapper, or what I understood about it anyway.

sdroege replied:
I was actually confused and thought `IOStreamAsyncReadWrite` could be created from the outside. That's fortunately not the case, so what you did before was actually correct. The wrapper is not needed (and if it was needed then the `new` function should've been unsafe: `GSocketConnection` is not thread-safe but it's safe to be sent to another thread... as long as all references to it are only used on the same thread).

sdroege replied:
Pffff, there actually is a constructor for it to build it from arbitrary `IOStream`s. I'm not sure why it's public but changing that is another breaking change, so you could go with a wrapper around it for `connect_async()` that in addition implements `Send` and proxies the trait impls...

And for 0.7 this constructor should probably be made private

sdroege commented:
The solution here would be to make the constructor private and release 0.7. Can you update the PR accordingly, I'd release 0.7 once enough other changes accumulated :)

For the time being this can be relatively easily worked around in application code.

philn committed:
gio: Make the IOStreamAsyncReadWrite constructor private

Ensuring this struct can only be created internally is important, ensuring
only one reference of GIOStream is tracked.

sdroege commented:
Manually merged
